src/Database/LSMTree/Internal.hs

```diff
@@ -1369,42 +1371,149 @@ unions ts = do
-
-          content <-
-            error "unions: combine contents into merging tree" $ -- TODO
-              contents
-
-          t <-
-            newWith
-              reg
-              sesh
-              seshEnv
-              conf
-              (error "unions: ArenaManager") -- TODO
-              content
-
+          t <- unionsInOpenSession sesh seshEnv reg ts
           pure (seshState, t)
 
+{-# SPECIALISE unionsInOpenSession ::
+     Session IO h
+  -> SessionEnv IO h
+  -> ActionRegistry IO
+  -> NonEmpty (Table IO h)
+  -> IO (Table IO h) #-}
+unionsInOpenSession ::
+     (MonadSTM m, MonadMask m, MonadMVar m, MonadST m)
+  => Session m h
+  -> SessionEnv m h
+  -> ActionRegistry m
+  -> NonEmpty (Table m h)
+  -> m (Table m h)
+unionsInOpenSession sesh seshEnv reg ts = do
+
+    -- The TableConfig for the new table is taken from the first
+    -- table in the union. This works because the new table is almost
+    -- completely fresh. It will have an empty write buffer and no
+    -- runs in the normal levels. All the existing runs get squashed
+    -- down into a single run before rejoining as a last level.
+    let conf = tableConfig (NE.head ts)
+
+    mts <- forM (NE.toList ts) $ \t ->
+      withOpenTable t $ \tEnv ->
+        RW.withReadAccess (tableContent tEnv) $ \tc -> do
+          -- tableContentToMergingTree duplicates all runs and merges
+          -- so the ones from the tableContent here do not escape
+          -- the read access.
+          mt <- withRollback reg
+                  (tableContentToMergingTree seshEnv conf reg tc)
+                  releaseRef
+          -- The mt here is a temporary value, since newPendingUnionMerge
+          -- will make its own references, so release mt at the end of
+          -- the action registry bracket
+          delayedCommit reg (releaseRef mt)
+          return mt
+    mt <- newPendingUnionMerge mts
+    empty <- newEmptyTableContent seshEnv reg
+    let content = empty { tableUnionLevel = Union mt }
+
+        -- Pick the arena manager to optimise the case of:
+        -- bigTableWithLotsOfLookups <> someUpdates
+        -- by reusing the arena manager from the first one.
+        am = tableArenaManager (NE.head ts)
+
+    newWith reg sesh seshEnv conf am content
+
+{-# SPECIALISE tableContentToMergingTree ::
+     SessionEnv IO h
+  -> TableConfig
+  -> ActionRegistry IO
+  -> TableContent IO h
+  -> IO (Ref (MergingTree IO h)) #-}
+tableContentToMergingTree ::
+     forall m h.
+     (MonadMask m, MonadMVar m, MonadST m, MonadSTM m)
+  => SessionEnv m h
+  -> TableConfig
+  -> ActionRegistry m
+  -> TableContent m h
+  -> m (Ref (MergingTree m h))
+tableContentToMergingTree seshEnv conf reg
+                          tc@TableContent {
+                            tableLevels,
+                            tableUnionLevel
+                          } =
+    bracket (writeBufferToNewRun seshEnv conf reg tc)
+            (mapM_ releaseRef) $ \mwbr ->
```

Comment:
This should use `delayedCommit`, for the double release reason I mention `writeBufferToNewRun`

Reply:
Alternatively, `writeBufferToNewRun` could not use the registry at all, right?

Reply:
That's probably also a valid alternative